# Deep Learning for Natural Language Processing with Pytorch
This tutorial will walk you through the key ideas of deep learning programming using Pytorch.
Many of the concepts (such as the computation graph abstraction and autograd) are not unique to Pytorch and are relevant to any deep learning tool kit out there.

I am writing this tutorial to focus specifically on NLP for people who have never written code in any deep learning framework (e.g, TensorFlow, Theano, Keras, Dynet).  It assumes working knowledge of core NLP problems: part-of-speech tagging, language modeling, etc.  It also assumes familiarity with neural networks at the level of an intro AI class (such as one from the Russel and Norvig book).  Usually, these courses cover the basic backpropagation algorithm on feed-forward neural networks, and make the point that they are chains of compositions of linearities and non-linearities.  This tutorial aims to get you started writing deep learning code, given you have this prerequisite knowledge.

Note this is about *models*, not data.  For all of the models, I just create a few test examples with small dimensionality so you can see how the weights change as it trains.  If you have some real data you want to try, you should be able to rip out any of the models from this notebook and use them on it.

In [1]:
import torch
import torch.autograd as autograd
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim

torch.manual_seed(1)

# 1. Introduction to Torch's tensor library

All of deep learning is computations on tensors, which are generalizations of a matrix that can be indexed in more than 2 dimensions.  We will see exactly what this means in-depth later.  First, lets look what we can do with tensors.

### Creating Tensors
Tensors can be created from Python lists with the torch.Tensor() function.

In [2]:
# Create a torch.Tensor object with the given data.  It is a 1D vector
V_data = [1., 2., 3.]
V = torch.Tensor(V_data)
print(V)

# Creates a matrix
M_data = [[1., 2., 3.], [4., 5., 6]]
M = torch.Tensor(M_data)
print(M)

# Create a 3D tensor of size 2x2x2.
T_data = [[[1.,2.], [3.,4.]],
          [[5.,6.], [7.,8.]]]
T = torch.Tensor(T_data)
print(T)

tensor([ 1.,  2.,  3.])
tensor([[ 1.,  2.,  3.],
        [ 4.,  5.,  6.]])
tensor([[[ 1.,  2.],
         [ 3.,  4.]],

        [[ 5.,  6.],
         [ 7.,  8.]]])


What is a 3D tensor anyway?
Think about it like this.
If you have a vector, indexing into the vector gives you a scalar.  If you have a matrix, indexing into the matrix gives you a vector.  If you have a 3D tensor, then indexing into the tensor gives you a matrix!

A note on terminology: when I say "tensor" in this tutorial, it refers to any torch.Tensor object.  Vectors and matrices are special cases of torch.Tensors, where their dimension is 1 and 2 respectively.  When I am talking about 3D tensors, I will explicitly use the term "3D tensor".

In [3]:
# Index into V and get a scalar
print(V[0])

# Index into M and get a vector
print(M[0])

# Index into T and get a matrix
print(T[0])

tensor(1.)
tensor([ 1.,  2.,  3.])
tensor([[ 1.,  2.],
        [ 3.,  4.]])


You can also create tensors of other datatypes.  The default, as you can see, is Float.
To create a tensor of integer types, try torch.LongTensor().  Check <a href="https://pytorch.org/docs/stable/tensors.html">the documentation</a> for more data types, but Float and Long will be the most common.

You can create a tensor with random data and the supplied dimensionality with torch.randn()

In [4]:
x = torch.randn((3, 4, 5))
print(x)

tensor([[[-1.5256, -0.7502, -0.6540, -1.6095, -0.1002],
         [-0.6092, -0.9798, -1.6091, -0.7121,  0.3037],
         [-0.7773, -0.2515, -0.2223,  1.6871,  0.2284],
         [ 0.4676, -0.6970, -1.1608,  0.6995,  0.1991]],

        [[ 0.8657,  0.2444, -0.6629,  0.8073,  1.1017],
         [-0.1759, -2.2456, -1.4465,  0.0612, -0.6177],
         [-0.7981, -0.1316,  1.8793, -0.0721,  0.1578],
         [-0.7735,  0.1991,  0.0457,  0.1530, -0.4757]],

        [[-0.1110,  0.2927, -0.1578, -0.0288,  0.4533],
         [ 1.1422,  0.2486, -1.7754, -0.0255, -1.0233],
         [-0.5962, -1.0055,  0.4285,  1.4761, -1.7869],
         [ 1.6103, -0.7040, -0.1853, -0.9962, -0.8313]]])


### Operations with Tensors
You can operate on tensors in the ways you would expect.

In [5]:
x = torch.Tensor([ 1., 2., 3. ])
y = torch.Tensor([ 4., 5., 6. ])
z = x + y
print(z)

tensor([ 5.,  7.,  9.])


See [the documentation](http://pytorch.org/docs/torch.html) for a complete list of the massive number of operations available to you.  They expand beyond just mathematical operations.

One helpful operation that we will make use of later is concatenation.

In [6]:
# By default, it concatenates along the first axis (concatenates rows)
x_1 = torch.randn(2, 5)
y_1 = torch.randn(3, 5)
z_1 = torch.cat([x_1, y_1])
print(z_1)

# Concatenate columns:
x_2 = torch.randn(2, 3)
y_2 = torch.randn(2, 5)
z_2 = torch.cat([x_2, y_2], 1) # second arg specifies which axis to concat along
print(z_2)

# If your tensors are not compatible, torch will complain. Uncomment to see the error
# torch.cat([x_1, x_2])

tensor([[-0.8029,  0.2366,  0.2857,  0.6898, -0.6331],
        [ 0.8795, -0.6842,  0.4533,  0.2912, -0.8317],
        [-0.5525,  0.6355, -0.3968, -0.6571, -1.6428],
        [ 0.9803, -0.0421, -0.8206,  0.3133, -1.1352],
        [ 0.3773, -0.2824, -2.5667, -1.4303,  0.5009]])
tensor([[ 0.5438, -0.4057,  1.1341, -0.1473,  0.6272,  1.0935,  0.0939,
          1.2381],
        [-1.1115,  0.3501, -0.7703, -1.3459,  0.5119, -0.6933, -0.1668,
         -0.9999]])


### Reshaping Tensors
Use the .view() method to reshape a tensor.
This method receives heavy use, because many neural network components expect their inputs to have a certain shape.
Often you will need to reshape before passing your data to the component.

In [7]:
x = torch.randn(2, 3, 4)
print(x)
print(x.view(2, 12)) # Reshape to 2 rows, 12 columns
print(x.view(2, -1)) # Same as above.  If one of the dimensions is -1, its size can be inferred

tensor([[[ 0.4175, -0.2127, -0.8400, -0.4200],
         [-0.6240, -0.9773,  0.8748,  0.9873],
         [-0.0594, -2.4919,  0.2423,  0.2883]],

        [[-0.1095,  0.3126,  1.5038,  0.5038],
         [ 0.6223, -0.4481, -0.2856,  0.3880],
         [-1.1435, -0.6512, -0.1032,  0.6937]]])
tensor([[ 0.4175, -0.2127, -0.8400, -0.4200, -0.6240, -0.9773,  0.8748,
          0.9873, -0.0594, -2.4919,  0.2423,  0.2883],
        [-0.1095,  0.3126,  1.5038,  0.5038,  0.6223, -0.4481, -0.2856,
          0.3880, -1.1435, -0.6512, -0.1032,  0.6937]])
tensor([[ 0.4175, -0.2127, -0.8400, -0.4200, -0.6240, -0.9773,  0.8748,
          0.9873, -0.0594, -2.4919,  0.2423,  0.2883],
        [-0.1095,  0.3126,  1.5038,  0.5038,  0.6223, -0.4481, -0.2856,
          0.3880, -1.1435, -0.6512, -0.1032,  0.6937]])



# 2. Computation Graphs and Automatic Differentiation

The concept of a computation graph is essential to efficient deep learning programming, because it allows you to not have to write the back propagation gradients yourself.  <font color=red>A computation graph is simply a specification of how your data is combined to give you the output.</font>  Since the graph totally specifies what parameters were involved with which operations, it contains enough information to compute derivatives.

__Before PyTorch 0.4.0:__ This probably sounds vague, so lets see what is going on using the fundamental class of Pytorch: autograd.Variable. <br>
__After PyTorch 0.4.0:__ This probably sounds vague, so let’s see what is going on using the flag `requires_grad`.

First, think from a programmers perspective.  What is stored in the torch.Tensor objects we were creating above?
Obviously the data and the shape, and maybe a few other things.  But when we added two tensors together, we got an output tensor.  All this output tensor knows is its data and shape.  It has no idea that it was the sum of two other tensors (it could have been read in from a file, it could be the result of some other operation, etc.)

__Before PyTorch 0.4.0:__ The Variable class with `requires_grad=True` keeps track of how it was created.  Lets see it in action. <br>
__After PyTorch 0.4.0:__ If `requires_grad=True`, the Tensor object keeps track of how it was created. Lets see it in action.

In [10]:
### Before PyTorch 0.4.0 ###
# Variables wrap tensor objects
x = autograd.Variable( torch.Tensor([1., 2., 3]), requires_grad=True )
# You can access the data with the .data attribute
print(x.data)

# You can also do all the same operations you did with tensors with Variables.
y = autograd.Variable( torch.Tensor([4., 5., 6]), requires_grad=True )
z = x + y
print(z.data)

# BUT z knows something extra.
print(z.grad_fn)

### After PyTorch 0.4.0 ###
# Tensor factory methods have a ``requires_grad`` flag
x = torch.tensor([1., 2., 3], requires_grad=True)

# With requires_grad=True, you can still do all the operations you previously
# could
y = torch.tensor([4., 5., 6], requires_grad=True)
z = x + y
print(z)

# BUT z knows something extra.
print(z.grad_fn)

tensor([ 1.,  2.,  3.])
tensor([ 5.,  7.,  9.])
tensor([ 5.,  7.,  9.])


So Variables/Tensors know what created them.  z knows that it wasn't read in from a file, it wasn't the result of a multiplication or exponential or whatever.  And if you keep following z.grad_fn, you will find yourself at x and y.

But how does that help us compute a gradient?

In [11]:
# Lets sum up all the entries in z
s = z.sum()
print(s)
print(s.grad_fn)

tensor(21.)


So now, what is the derivative of this sum with respect to the first component of x?  In math, we want
$$ \frac{\partial s}{\partial x_0} $$
Well, s knows that it was created as a sum of the tensor z.  z knows that it was the sum x + y.
So 
$$ s = \overbrace{x_0 + y_0}^\text{$z_0$} + \overbrace{x_1 + y_1}^\text{$z_1$} + \overbrace{x_2 + y_2}^\text{$z_2$} $$
And so s contains enough information to determine that the derivative we want is 1!

Of course this glosses over the challenge of how to actually compute that derivative.  The point here is that s is carrying along enough information that it is possible to compute it.  In reality, the developers of Pytorch program the sum() and + operations to know how to compute their gradients, and run the back propagation algorithm.  An in-depth discussion of that algorithm is beyond the scope of this tutorial.

Let's have Pytorch compute the gradient, and see that we were right: (note if you run this block multiple times, the gradient will increment (For instance I ran it twice :D).  That is because Pytorch *accumulates* the gradient into the .grad property, since for many models this is very convenient.)

In [13]:
# ss = 2 * s
# print(ss)
# ss.backward()

s.backward() # calling .backward() on any variable will run backprop, starting from it.
print(x.grad)

tensor([ 2.,  2.,  2.])


Understanding what is going on in the block below is crucial for being a successful programmer in deep learning.

In [14]:
### Before PyTorch 0.4.0 ###
print("### Before PyTorch 0.4.0 ###")
x = torch.randn((2,2))
y = torch.randn((2,2))
z = x + y # These are Tensor types, and backprop would not be possible
print(z)

var_x = autograd.Variable( x, requires_grad=True )
var_y = autograd.Variable( y, requires_grad=True )
var_z = var_x + var_y # var_z contains enough information to compute gradients, as we saw above (though requires_grad is crucial)
# print(var_z)
# print(var_z.data)
print(var_z.grad_fn)

var_z_data = var_z.data # Get the wrapped Tensor object out of var_z...
new_var_z = autograd.Variable( var_z_data ) # Re-wrap the tensor in a new variable

# ... does new_var_z have information to backprop to x and y?
# NO!
print(new_var_z.grad_fn)
# And how could it?  We yanked the tensor out of var_z (that is what var_z.data is).  This tensor
# doesn't know anything about how it was computed.  We pass it into new_var_z, and this is all the information
# new_var_z gets.  If var_z_data doesn't know how it was computed, theres no way new_var_z will.
# In essence, we have broken the variable away from its past history

print()

### After PyTorch 0.4.0 ###
print("### After PyTorch 0.4.0 ###")
x = torch.randn(2, 2)
y = torch.randn(2, 2)
# By default, user created Tensors have ``requires_grad=False``
print(x.requires_grad, y.requires_grad)
z = x + y
# So you can't backprop through z
print(z.grad_fn)

# ``.requires_grad_( ... )`` changes an existing Tensor's ``requires_grad``
# flag in-place. The input flag defaults to ``True`` if not given.
x = x.requires_grad_()
y = y.requires_grad_()
# z contains enough information to compute gradients, as we saw above
z = x + y
print(z.grad_fn)
# If any input to an operation has ``requires_grad=True``, so will the output
print(z.requires_grad)

# Now z has the computation history that relates itself to x and y
# Can we just take its values, and **detach** it from its history?
new_z = z.detach()

# ... does new_z have information to backprop to x and y?
# NO!
print(new_z.grad_fn)
# And how could it? ``z.detach()`` returns a tensor that shares the same storage
# as ``z``, but with the computation history forgotten. It doesn't know anything
# about how it was computed.
# In essence, we have broken the Tensor away from its past history

### Before PyTorch 0.4.0 ###
tensor([[-0.8245,  0.4056],
        [-1.4979, -2.2299]])
None

### After PyTorch 0.4.0 ###
False False
None
True
None


You can also stop autograd from tracking history on Tensors
with `requires_grad=True` by wrapping the code block in
`with torch.no_grad():`

In [16]:
x = torch.tensor([1., 2., 4.], requires_grad=True)
print(x.requires_grad)
print((x ** 2).requires_grad)

with torch.no_grad():
    print((x ** 2).requires_grad)

True
True
False


Here is the basic, extremely important rule for computing with autograd.Variables/Tensors (note this is more general than Pytorch.  There is an equivalent object in every major deep learning toolkit):

** If you want the error from your loss function to backpropogate to a component of your network, you MUST NOT break the Variable chain from that component to your loss Variable.  If you do, the loss will have no idea your component exists, and its parameters can't be updated. **

I say this in bold, because this error can creep up on you in very subtle ways (I will show some such ways below), and it will not cause your code to crash or complain, so you must be careful.

# 3. Deep Learning Building Blocks: Affine maps, non-linearities and objectives

Deep learning consists of composing linearities with non-linearities in clever ways.  The introduction of non-linearities allows for powerful models.  In this section, we will play with these core components, make up an objective function, and see how the model is trained.

### Affine Maps
One of the core workhorses of deep learning is the affine map, which is a function $f(x)$ where
$$ f(x) = Ax + b $$ for a matrix $A$ and vectors $x, b$.  The parameters to be learned here are $A$ and $b$.  Often, $b$ is refered to as the *bias* term.

<font color=red>Pytorch and most other deep learning frameworks do things a little differently than traditional linear algebra.  It maps the rows of the input instead of the columns.  That is, the $i$'th row of the output below is the mapping of the $i$'th row of the input under $A$, plus the bias term.</font>  Look at the example below.

In [17]:
lin = nn.Linear(5, 3) # maps from R^5 to R^3, parameters A (3*5), b (3)
data = autograd.Variable( torch.randn(2, 5) ) # data is 2x5.  A maps from 5 to 3... can we map "data" under A?
print(lin(data)) # yes

tensor([[-0.1176,  0.0377,  0.4498],
        [-0.5548,  0.5594,  0.2233]])


### Non-Linearities
First, note the following fact, which will explain why we need non-linearities in the first place.
Suppose we have two affine maps $f(x) = Ax + b$ and $g(x) = Cx + d$.  What is $f(g(x))$?
$$ f(g(x)) = A(Cx + d) + b = ACx + (Ad + b) $$
$AC$ is a matrix and $Ad + b$ is a vector, so we see that composing affine maps gives you an affine map.

From this, you can see that if you wanted your neural network to be long chains of affine compositions, that this adds no new power to your model than just doing a single affine map.

If we introduce non-linearities in between the affine layers, this is no longer the case, and we can build much more powerful models.

There are a few core non-linearities.  $\tanh(x), \sigma(x), \text{ReLU}(x)$ are the most common.
You are probably wondering: "why these functions?  I can think of plenty of other non-linearities."
The reason for this is that they have gradients that are easy to compute, and computing gradients is essential for learning.  For example
$$ \frac{d\sigma}{dx} = \sigma(x)(1 - \sigma(x)) $$

A quick note: although you may have learned some neural networks in your intro to AI class where $\sigma(x)$ was the default non-linearity, typically people shy away from it in practice.  This is because the gradient *vanishes* very quickly as the absolute value of the argument grows.  Small gradients means it is hard to learn.  Most people default to tanh or ReLU.

In [18]:
# In pytorch, most non-linearities are in torch.functional (we have it imported as F)
# Note that non-linearites typically don't have parameters like affine maps do.
# That is, they don't have weights that are updated during training.
data = torch.randn(2, 2)
print(data)
print(F.relu(data))

tensor([[ 1.4330,  1.6689],
        [ 1.8068, -0.6527]])
tensor([[ 1.4330,  1.6689],
        [ 1.8068,  0.0000]])


### Softmax and Probabilities
The function $\text{Softmax}(x)$ is also just a non-linearity, but it is special in that it usually is the last operation done in a network.  This is because it takes in a vector of real numbers and returns a probability distribution.  Its definition is as follows.  Let $x$ be a vector of real numbers (positive, negative, whatever, there are no constraints).  Then the i'th component of $\text{Softmax}(x)$ is
$$ \frac{\exp(x_i)}{\sum_j \exp(x_j)} $$
It should be clear that the output is a probability distribution: each element is non-negative and the sum over all components is 1.

You could also think of it as just applying an element-wise exponentiation operator to the input to make everything non-negative and then dividing by the normalization constant.

In [19]:
# Softmax is also in torch.functional
data = torch.randn(5)
print(data)
print(F.softmax(data, dim=0))
print(F.softmax(data, dim=0).sum()) # Sums to 1 because it is a distribution!
print(F.log_softmax(data, dim=0)) # theres also log_softmax

tensor([ 1.0488,  0.4975,  0.3865, -1.5278,  0.3892])
tensor([ 0.3724,  0.2146,  0.1921,  0.0283,  0.1926])
tensor(1.0000)
tensor([-0.9877, -1.5389, -1.6499, -3.5643, -1.6473])


### Objective Functions
The objective function is the function that your network is being trained to minimize (in which case it is often called a *loss function* or *cost function*).
This proceeds by first choosing a training instance, running it through your neural network, and then computing the loss of the output.  The parameters of the model are then updated by taking the derivative of the loss function.  Intuitively, if your model is completely confident in its answer, and its answer is wrong, your loss will be high.  If it is very confident in its answer, and its answer is correct, the loss will be low.

The idea behind minimizing the loss function on your training examples is that your network will hopefully generalize well and have small loss on unseen examples in your dev set, test set, or in production.
An example loss function is the *negative log likelihood loss*, which is a very common objective for multi-class classification.  For supervised multi-class classification, this means training the network to minimize the negative log probability of the correct output (or equivalently, maximize the log probability of the correct output).

# 4. Optimization and Training

So what we can compute a loss function for an instance?  What do we do with that?
We saw earlier that autograd.Variable's know how to compute gradients with respect to the things that were used to compute it.  Well, since our loss is an autograd.Variable, we can compute gradients with respect to all of the parameters used to compute it!  Then we can perform standard gradient updates.  Let $\theta$ be our parameters, $L(\theta)$ the loss function, and $\eta$ a positive learning rate.  Then:

$$ \theta^{(t+1)} = \theta^{(t)} - \eta \nabla_\theta L(\theta) $$

There are a huge collection of algorithms and active research in attempting to do something more than just this vanilla gradient update.  Many attempt to vary the learning rate based on what is happening at train time.  You don't need to worry about what specifically these algorithms are doing unless you are really interested.  Torch provies many in the torch.optim package, and they are all completely transparent.  Using the simplest gradient update is the same as the more complicated algorithms.  Trying different update algorithms and different parameters for the update algorithms (like different initial learning rates) is important in optimizing your network's performance.  Often, just replacing vanilla SGD with an optimizer like Adam or RMSProp will boost performance noticably.

# 5. Creating Network Components in Pytorch
Before we move on to our focus on NLP, lets do an annotated example of building a network in Pytorch using only affine maps and non-linearities.  We will also see how to compute a loss function, using Pytorch's built in negative log likelihood, and update parameters by backpropagation.

<font color=red>All network components should inherit from nn.Module and override the forward() method.</font>  That is about it, as far as the boilerplate is concerned.  Inheriting from nn.Module provides functionality to your component.  For example, it makes it keep track of its trainable parameters, you can swap it between CPU and GPU with the .cuda() or .cpu() functions, etc.

Let's write an annotated example of a network that takes in a sparse bag-of-words (BOW) representation and outputs a probability distribution over two labels: "English" and "Spanish".  This model is just logistic regression.

### Example: Logistic Regression Bag-of-Words classifier
Our model will map a sparse BOW representation to log probabilities over labels.  We assign each word in the vocab an index.  For example, say our entire vocab is two words "hello" and "world", with indices 0 and 1 respectively.
The BoW vector for the sentence "hello hello hello hello" is
$$ \left[ 4, 0 \right] $$
For "hello world world hello", it is 
$$ \left[ 2, 2 \right] $$
etc.
In general, it is
$$ \left[ \text{Count}(\text{hello}), \text{Count}(\text{world}) \right] $$

Denote this BOW vector as $x$.
The output of our network is:
$$ \log \text{Softmax}(Ax + b) $$
That is, we pass the input through an affine map and then do log softmax.

In [20]:
data = [ ("me gusta comer en la cafeteria".split(), "SPANISH"),
         ("Give it to me".split(), "ENGLISH"),
         ("No creo que sea una buena idea".split(), "SPANISH"),
         ("No it is not a good idea to get lost at sea".split(), "ENGLISH") ]

test_data = [ ("Yo creo que si".split(), "SPANISH"),
              ("it is lost on me".split(), "ENGLISH")]

# word_to_ix maps each word in the vocab to a unique integer, which will be its
# index into the Bag of words vector
word_to_ix = {}
for sent, _ in data + test_data:
    for word in sent:
        if word not in word_to_ix:
            word_to_ix[word] = len(word_to_ix)
print(word_to_ix)

VOCAB_SIZE = len(word_to_ix)
NUM_LABELS = 2

{'en': 3, 'to': 8, 'on': 25, 'comer': 2, 'sea': 12, 'buena': 14, 'que': 11, 'No': 9, 'creo': 10, 'me': 0, 'una': 13, 'lost': 21, 'a': 18, 'Give': 6, 'gusta': 1, 'cafeteria': 5, 'get': 20, 'Yo': 23, 'la': 4, 'it': 7, 'good': 19, 'idea': 15, 'not': 17, 'is': 16, 'si': 24, 'at': 22}


In [21]:
class BoWClassifier(nn.Module): # inheriting from nn.Module!
    
    def __init__(self, num_labels, vocab_size):
        # calls the init function of nn.Module.  Dont get confused by syntax,
        # just always do it in an nn.Module
        super(BoWClassifier, self).__init__()
        
        # Define the parameters that you will need.  In this case, we need A and b,
        # the parameters of the affine mapping.
        # Torch defines nn.Linear(), which provides the affine map.
        # Make sure you understand why the input dimension is vocab_size
        # and the output is num_labels!
        self.linear = nn.Linear(vocab_size, num_labels)
        
        # NOTE! The non-linearity log softmax does not have parameters! So we don't need
        # to worry about that here
        
    # Don't forget to override forward()!
    def forward(self, bow_vec):
        # Pass the input through the linear layer,
        # then pass that through log_softmax.
        # Many non-linearities and other functions are in torch.nn.functional
        return F.log_softmax(self.linear(bow_vec), dim=1)

In [22]:
def make_bow_vector(sentence, word_to_ix):
    vec = torch.zeros(len(word_to_ix))
    for word in sentence:
        vec[word_to_ix[word]] += 1
    # print(vec)
    # print(vec.view(1, -1))
    return vec.view(1, -1)

def make_target(label, label_to_ix):
    return torch.LongTensor([label_to_ix[label]])

In [23]:
model = BoWClassifier(NUM_LABELS, VOCAB_SIZE)

# The model knows its parameters.  The first output below is A, the second is b.
# Whenever you assign a component (nn.Linear(...) in our case) to a class variable (self.linear in our case)
# in the __init__ function of a module,
# which was done with the line
# self.linear = nn.Linear(...),
# then through some Python magic from the PyTorch devs, your module (in this case, BoWClassifier)
# will store knowledge of the nn.Linear's parameters
for param in model.parameters():
    print(param)

Parameter containing:
tensor([[-0.1710,  0.1650, -0.0372,  0.0396,  0.0073, -0.1250,  0.1104,
          0.1099,  0.0099, -0.1115, -0.0833,  0.0027, -0.1120, -0.1094,
         -0.0293, -0.0565,  0.0481, -0.0515, -0.0260, -0.0749, -0.1792,
          0.1710,  0.0374,  0.1754, -0.0316, -0.0493],
        [-0.1844, -0.0744,  0.1286, -0.1921, -0.0686,  0.1195,  0.1130,
          0.0724, -0.0388, -0.0148, -0.0372, -0.0723,  0.0818, -0.0668,
         -0.1102,  0.0445, -0.1418, -0.0419,  0.1002,  0.0733,  0.1670,
         -0.1338,  0.0017, -0.0579, -0.1097, -0.1103]])
Parameter containing:
tensor(1.00000e-02 *
       [ 4.9409,  2.0519])


In [24]:
# To run the model, pass in a BoW vector
sample = data[0]
bow_vector = make_bow_vector(sample[0], word_to_ix)
log_probs = model(bow_vector)
# log_probs = model(autograd.Variable(bow_vector))
print(log_probs)

tensor([[-0.6077, -0.7866]])


Which of the above values corresponds to the log probability of ENGLISH, and which to SPANISH?  We never defined it, but we need to if we want to train the thing.

In [25]:
label_to_ix = { "SPANISH": 0, "ENGLISH": 1 }

So lets train!  To do this, we pass instances through to get log probabilities, compute a loss function, compute the gradient of the loss function, and then update the parameters with a gradient step.  Loss functions are provided by Torch in the nn package.  nn.NLLLoss() is the negative log likelihood loss we want.  It also defines optimization functions in torch.optim.  Here, we will just use SGD.

<font color=red>Note that the *input* to NLLLoss is a vector of log probabilities, and a target label.</font> It doesn't compute the log probabilities for us.  This is why the last layer of our network is log softmax.
The loss function nn.CrossEntropyLoss() is the same as NLLLoss(), except it does the log softmax for you.

In [26]:
# Run on test data before we train, just to see a before-and-after
for instance, label in test_data:
    bow_vec = make_bow_vector(instance, word_to_ix)
    log_probs = model(bow_vec)
    print(log_probs)
print(next(model.parameters())[:,word_to_ix["creo"]]) # Print the matrix column corresponding to "creo"

tensor([[-0.5255, -0.8947]])
tensor([[-0.4251, -1.0605]])
tensor(1.00000e-02 *
       [-8.3331, -3.7244])


In [27]:
loss_function = nn.NLLLoss()

# To construct an Optimizer you have to give it an iterable containing the parameters (all should be Variables)
# to optimize. Then, you can specify optimizer-specific options such as the learning rate, weight decay, etc.
optimizer = optim.SGD(model.parameters(), lr=0.1)

# Usually you want to pass over the training data several times.
# 100 is much bigger than on a real data set, but real datasets have more than
# two instances.  Usually, somewhere between 5 and 30 epochs is reasonable.
for epoch in range(100):
    for instance, label in data:
        # Step 1. Remember that Pytorch accumulates gradients.  We need to clear them out
        # before each instance
        # Once run with commenting the following line out and see what happens
        model.zero_grad()
    
        # Step 2. Make our BOW vector and also we must wrap the target in a Variable
        # as an integer.  For example, if the target is SPANISH, then we wrap the integer
        # 0.  The loss function then knows that the 0th element of the log probabilities is
        # the log probability corresponding to SPANISH
        bow_vec = make_bow_vector(instance, word_to_ix)
        target = make_target(label, label_to_ix)
    
        # Step 3. Run our forward pass.
        log_probs = model(bow_vec)
    
        # Step 4. Compute the loss, gradients, and update the parameters by calling
        # optimizer.step()
        loss = loss_function(log_probs, target)
        loss.backward()
        optimizer.step()

    print("Finished epoch {}, with loss: {}".format(epoch, loss))

Finished epoch 0, with loss: 0.6128547191619873
Finished epoch 1, with loss: 0.2654516398906708
Finished epoch 2, with loss: 0.16969987750053406
Finished epoch 3, with loss: 0.12449388206005096
Finished epoch 4, with loss: 0.09803160279989243
Finished epoch 5, with loss: 0.08067663758993149
Finished epoch 6, with loss: 0.06844251602888107
Finished epoch 7, with loss: 0.05937039852142334
Finished epoch 8, with loss: 0.05238448455929756
Finished epoch 9, with loss: 0.046845268458127975
Finished epoch 10, with loss: 0.042349182069301605
Finished epoch 11, with loss: 0.03862926736474037
Finished epoch 12, with loss: 0.03550203889608383
Finished epoch 13, with loss: 0.03283732384443283
Finished epoch 14, with loss: 0.030540313571691513
Finished epoch 15, with loss: 0.028540320694446564
Finished epoch 16, with loss: 0.026783594861626625
Finished epoch 17, with loss: 0.02522856928408146
Finished epoch 18, with loss: 0.02384260483086109
Finished epoch 19, with loss: 0.022599685937166214
Finish

In [28]:
for instance, label in test_data:
    bow_vec = make_bow_vector(instance, word_to_ix)
    log_probs = model(bow_vec)
    print(log_probs)
print(next(model.parameters())[:,word_to_ix["creo"]]) # Index corresponding to Spanish goes up, English goes down!

tensor([[-0.0927, -2.4247]])
tensor([[-2.1196, -0.1279]])
tensor([ 0.3637, -0.4843])


We got the right answer!  You can see that the log probability for Spanish is much higher in the first example, and the log probability for English is much higher in the second for the test data, as it should be.

Now you see how to make a Pytorch component, pass some data through it and do gradient updates.
We are ready to dig deeper into what deep NLP has to offer.

# 6. Word Embeddings: Encoding Lexical Semantics

Word embeddings are dense vectors of real numbers, one per word in your vocabulary.
In NLP, it is almost always the case that your features are words!  But how should you represent a word in a computer?
You could store its ascii character representation, but that only tells you what the word *is*, it doesn't say much about what it *means* (you might be able to derive its part of speech from its affixes, or properties from its capitalization, but not much).  Even more, in what sense could you combine these representations?
We often want dense outputs from our neural networks, where the inputs are $|V|$ dimensional, where $V$ is our vocabulary, but often the outputs are only a few dimensional (if we are only predicting a handful of labels, for instance).  How do we get from a massive dimensional space to a smaller dimensional space?

How about instead of ascii representations, we use a one-hot encoding?  That is, we represent the word $w$ by
$$ \overbrace{\left[ 0, 0, \dots, 1, \dots, 0, 0 \right]}^\text{|V| elements} $$
where the 1 is in a location unique to $w$.  Any other word will have a 1 in some other location, and a 0 everywhere else.

There is an enormous drawback to this representation, besides just how huge it is.  It basically treats all words as independent entities with no relation to each other.  What we really want is some notion of *similarity* between words.  Why?  Let's see an example.

Suppose we are building a language model.  Suppose we have seen the sentences
* The mathematician ran to the store.
* The physicist ran to the store.
* The mathematician solved the open problem.

in our training data.
Now suppose we get a new sentence never before seen in our training data:
* The physicist solved the open problem.

Our language model might do OK on this sentence, but wouldn't it be much better if we could use the following two facts:
* We have seen mathematician and physicist in the same role in a sentence.  Somehow they have a semantic relation.
* We have seen mathematician in the same role in this new unseen sentence as we are now seeing physicist.

and then infer that physicist is actually a good fit in the new unseen sentence?  This is what we mean by a notion of similarity: we mean *semantic similarity*, not simply having similar orthographic representations.  It is a technique to combat the sparsity of linguistic data, by connecting the dots between what we have seen and what we haven't.  This example of course relies on a fundamental linguistic assumption: that words appearing in similar contexts are related to each other semantically.  This is called the [distributional hypothesis](https://en.wikipedia.org/wiki/Distributional_semantics). For more details, refer to <a href="https://web.stanford.edu/~jurafsky/slp3/15.pdf">chapter 15</a> and <a href="https://web.stanford.edu/~jurafsky/slp3/16.pdf">chapter 16</a> of __Speech and Language Processing by Dan Jurafsky and James H. Martin__.

### Getting Dense Word Embeddings

How can we solve this problem?  That is, how could we actually encode semantic similarity in words?
Maybe we think up some semantic attributes.  For example, we see that both mathematicians and physicists can run, so maybe we give these words a high score for the "is able to run" semantic attribute.  Think of some other attributes, and imagine what you might score some common words on those attributes.

If each attribute is a dimension, then we might give each word a vector, like this:
$$ q_\text{mathematician} = \left[ \overbrace{2.3}^\text{can run},
\overbrace{9.4}^\text{likes coffee}, \overbrace{-5.5}^\text{majored in Physics}, \dots \right] $$
$$ q_\text{physicist} = \left[ \overbrace{2.5}^\text{can run},
\overbrace{9.1}^\text{likes coffee}, \overbrace{6.4}^\text{majored in Physics}, \dots \right] $$

Then we can get a measure of similarity between these words by doing:
$$ \text{Similarity}(\text{physicist}, \text{mathematician}) = q_\text{physicist} \cdot q_\text{mathematician} $$

Although it is more common to normalize by the lengths:
$$ \text{Similarity}(\text{physicist}, \text{mathematician}) = \frac{q_\text{physicist} \cdot q_\text{mathematician}}
{\| q_\text{\physicist} \| \| q_\text{mathematician} \|} = \cos (\phi) $$
Where $\phi$ is the angle between the two vectors.  That way, extremely similar words (words whose embeddings point in the same direction) will have similarity 1.  Extremely dissimilar words should have similarity -1.

You can think of the sparse one-hot vectors from the beginning of this section as a special case of these new vectors we have defined, where each word basically has similarity 0, and we gave each word some unique semantic attribute.  These new vectors are *dense*, which is to say their entries are (typically) non-zero.

But these new vectors are a big pain: you could think of thousands of different semantic attributes that might be relevant to determining similarity, and how on earth would you set the values of the different attributes?  Central to the idea of deep learning is that the neural network learns representations of the features, rather than requiring the programmer to design them herself.  So why not just let the word embeddings be parameters in our model, and then be updated during training?  This is exactly what we will do.  <font color=red>We will have some *latent semantic attributes* that the network can, in principle, learn.</font>  Note that the word embeddings will probably not be interpretable.  That is, although with our hand-crafted vectors above we can see that mathematicians and physicists are similar in that they both like coffee, if we allow a neural network to learn the embeddings and see that both mathematicians and physicists have a large value in the second dimension, it is not clear what that means.  They are similar in some latent semantic dimension, but this probably has no interpretation to us.

In summary, **word embeddings are a representation of the *semantics* of a word, efficiently encoding semantic information that might be relevant to the task at hand**.  You can embed other things too: part of speech tags, parse trees, anything!  The idea of feature embeddings is central to the field.

### Word Embeddings in Pytorch
Before we get to a worked example and an exercise, a few quick notes about how to use embeddings in Pytorch and in deep learning programming in general.
Similar to how we defined a unique index for each word when making one-hot vectors, we also need to define an index for each word when using embeddings.  These will be keys into a lookup table.  That is, embeddings are stored as a $|V| \times D$ matrix, where $D$ is the dimensionality of the embeddings, such that the word assigned index $i$ has its embedding stored in the $i$'th row of the matrix.  In all of my code, the mapping from words to indices is a dictionary named word_to_ix.

The module that allows you to use embeddings is torch.nn.Embedding, which takes two arguments: the vocabulary size, and the dimensionality of the embeddings.

To index into this table, you must use torch.LongTensor (since the indices are integers, not floats).

In [29]:
word_to_ix = { "hello": 0, "world": 1 }
embeds = nn.Embedding(2, 5) # 2 words in vocab, 5 dimensional embeddings
lookup_tensor = torch.LongTensor([word_to_ix["hello"]])
# hello_embed = embeds( autograd.Variable(lookup_tensor) )
hello_embed = embeds(lookup_tensor)
print(hello_embed)

tensor([[ 1.4697, -0.3951, -0.5101,  1.1163, -0.5926]])


### An Example: N-Gram Language Modeling
Recall that in an n-gram language model, given a sequence of words $w$, we want to compute
$$ P(w_i | w_{i-1}, w_{i-2}, \dots, w_{i-n+1} ) $$
Where $w_i$ is the ith word of the sequence.

In this example, we will compute the loss function on some training examples and update the parameters with backpropagation.

In [30]:
CONTEXT_SIZE = 2
EMBEDDING_DIM = 10
# We will use Shakespeare Sonnet 2
test_sentence = """When forty winters shall besiege thy brow,
And dig deep trenches in thy beauty's field,
Thy youth's proud livery so gazed on now,
Will be a totter'd weed of small worth held:
Then being asked, where all thy beauty lies,
Where all the treasure of thy lusty days;
To say, within thine own deep sunken eyes,
Were an all-eating shame, and thriftless praise.
How much more praise deserv'd thy beauty's use,
If thou couldst answer 'This fair child of mine
Shall sum my count, and make my old excuse,'
Proving his beauty by succession thine!
This were to be new made when thou art old,
And see thy blood warm when thou feel'st it cold.""".split()
# we should tokenize the input, but we will ignore that for now
# build a list of tuples.  Each tuple is ([ word_i-2, word_i-1 ], target word)
trigrams = [ ([test_sentence[i], test_sentence[i+1]], test_sentence[i+2]) for i in range(len(test_sentence) - 2) ]
print(trigrams[:3]) # print the first 3, just so you can see what they look like

[(['When', 'forty'], 'winters'), (['forty', 'winters'], 'shall'), (['winters', 'shall'], 'besiege')]


In [31]:
vocab = set(test_sentence)
word_to_ix = { word: i for i, word in enumerate(vocab) }

In [32]:
class NGramLanguageModeler(nn.Module):
    
    def __init__(self, vocab_size, embedding_dim, context_size):
        super(NGramLanguageModeler, self).__init__()
        
        # Model that "we" designed:
        # input context -> embeddings
        # embeddings -> 128-node hidden state
        # 128-node hidden state -> output
        self.embeddings = nn.Embedding(vocab_size, embedding_dim)
        self.linear1 = nn.Linear(context_size * embedding_dim, 128)
        self.linear2 = nn.Linear(128, vocab_size)

        # Again, note that nonlinearities do not have trainable parameters, so they don't appear here!
        
    def forward(self, inputs):
        embeds = self.embeddings(inputs).view((1, -1))
        out = F.relu(self.linear1(embeds))
        out = self.linear2(out)
        log_probs = F.log_softmax(out, dim=1)
        return log_probs

In [33]:
losses = []
loss_function = nn.NLLLoss()
model = NGramLanguageModeler(len(vocab), EMBEDDING_DIM, CONTEXT_SIZE)
optimizer = optim.SGD(model.parameters(), lr=0.001)

for epoch in range(20):
    total_loss = torch.Tensor([0])
    for context, target in trigrams:
    
        # Step 1. Prepare the inputs to be passed to the model (i.e, turn the words
        # into integer indices and wrap them in variables)
        context_idxs = list(map(lambda w: word_to_ix[w], context))
        context_var = torch.LongTensor(context_idxs)
    
        # Step 2. Recall that torch *accumulates* gradients.  Before passing in a new instance,
        # you need to zero out the gradients from the old instance
        model.zero_grad()
    
        # Step 3. Run the forward pass, getting log probabilities over next words
        log_probs = model(context_var)
    
        # Step 4. Compute your loss function. (Again, Torch wants the target word wrapped in a variable)
        loss = loss_function(log_probs, torch.LongTensor([word_to_ix[target]]))
    
        # Step 5. Do the backward pass and update the gradient
        loss.backward()
        optimizer.step()
    
        total_loss += loss.data
    losses.append(total_loss)
print(losses) # The loss decreased every iteration over the training data!

[tensor([ 518.9597]), tensor([ 516.7561]), tensor([ 514.5637]), tensor([ 512.3822]), tensor([ 510.2117]), tensor([ 508.0527]), tensor([ 505.9041]), tensor([ 503.7639]), tensor([ 501.6310]), tensor([ 499.5036]), tensor([ 497.3819]), tensor([ 495.2652]), tensor([ 493.1532]), tensor([ 491.0457]), tensor([ 488.9426]), tensor([ 486.8437]), tensor([ 484.7476]), tensor([ 482.6539]), tensor([ 480.5614]), tensor([ 478.4708])]


### Exercise: Computing Word Embeddings: Continuous Bag-of-Words

The Continuous Bag-of-Words model (CBOW) is frequently used in NLP deep learning.  It is a model that tries to predict words given the context of a few words before and a few words after the target word.  This is distinct from language modeling, since CBOW is not sequential and does not have to be probabilistic.  Typcially, CBOW is used to quickly train word embeddings, and these embeddings are used to initialize the embeddings of some more complicated model.  Usually, this is referred to as *pretraining embeddings*.  It almost always helps performance a couple of percent.

The CBOW model is as follows.  Given a target word $w_i$ and an $N$ context window on each side, $w_{i-1}, \dots, w_{i-N}$ and $w_{i+1}, \dots, w_{i+N}$, referring to all context words collectively as $C$, CBOW tries to minimize
$$ -\log p(w_i | C) = \log \text{Softmax}(A(\sum_{w \in C} q_w) + b) $$
where $q_w$ is the embedding for word $w$.

Implement this model in Pytorch by filling in the class below. Some tips:
* Think about which parameters you need to define.
* Make sure you know what shape each operation expects.  Use .view() if you need to reshape.

In [34]:
CONTEXT_SIZE = 2 # 2 words to the left, 2 to the right
raw_text = """We are about to study the idea of a computational process. Computational processes are abstract
beings that inhabit computers. As they evolve, processes manipulate other abstract
things called data. The evolution of a process is directed by a pattern of rules
called a program. People create programs to direct processes. In effect,
we conjure the spirits of the computer with our spells.""".split()
word_to_ix = { word: i for i, word in enumerate(set(raw_text)) }
data = []
for i in range(2, len(raw_text) - 2):
    context = [ raw_text[i-2], raw_text[i-1], raw_text[i+1], raw_text[i+2] ]
    target = raw_text[i]
    data.append( (context, target) )
print(data[:5])

[(['We', 'are', 'to', 'study'], 'about'), (['are', 'about', 'study', 'the'], 'to'), (['about', 'to', 'the', 'idea'], 'study'), (['to', 'study', 'idea', 'of'], 'the'), (['study', 'the', 'of', 'a'], 'idea')]


In [35]:
class CBOW(nn.Module):
    
    def __init__(self, vocab_size, embedding_dim):
        # pass
        super(CBOW, self).__init__()
        
        self.embeddings = nn.Embedding(vocab_size, embedding_dim)
        self.linear = nn.Linear(embedding_dim, vocab_size)
    
    def forward(self, inputs):
        # pass
        embeds = self.embeddings(inputs)
        out = embeds.sum(dim=0).view(1, -1)
        log_probs = F.log_softmax(self.linear(out), dim=1)
        return log_probs

In [36]:
# create your model and train.  here are some functions to help you make the data ready for use by your module
def make_context_vector(context, word_to_ix):
    idxs = list(map(lambda w: word_to_ix[w], context))
    tensor = torch.LongTensor(idxs)
    return tensor

losses = []
loss_function = nn.NLLLoss()
model = CBOW(len(set(raw_text)), 10)
optimizer = optim.SGD(model.parameters(), lr=0.001)

for epoch in range(20):
    total_loss = torch.Tensor([0])
    for context, target in data:

        # Step 1. Prepare the inputs to be passed to the model (i.e, turn the words
        # into integer indices and wrap them in variables)
        context_idxs = make_context_vector(context, word_to_ix)

        # Step 2. Recall that torch *accumulates* gradients. Before passing in a
        # new instance, you need to zero out the gradients from the old
        # instance
        model.zero_grad()

        # Step 3. Run the forward pass, getting log probabilities over next
        # words
        log_probs = model(context_idxs)

        # Step 4. Compute your loss function. (Again, Torch wants the target
        # word wrapped in a variable)
        loss = loss_function(log_probs, torch.tensor([word_to_ix[target]], dtype=torch.long))

        # Step 5. Do the backward pass and update the gradient
        loss.backward()
        optimizer.step()

        # Get the Python number from a 1-element Tensor by calling tensor.item()
        total_loss += loss.item()
    losses.append(total_loss)
print(losses)  # The loss decreased every iteration over the training data!

[tensor([ 266.8810]), tensor([ 263.7825]), tensor([ 260.7372]), tensor([ 257.7432]), tensor([ 254.7985]), tensor([ 251.9017]), tensor([ 249.0515]), tensor([ 246.2465]), tensor([ 243.4853]), tensor([ 240.7672]), tensor([ 238.0911]), tensor([ 235.4560]), tensor([ 232.8612]), tensor([ 230.3059]), tensor([ 227.7895]), tensor([ 225.3112]), tensor([ 222.8706]), tensor([ 220.4669]), tensor([ 218.0999]), tensor([ 215.7690])]


# 7. Sequence Models and Long-Short Term Memory Networks

At this point, we have seen various feed-forward networks.
That is, there is no state maintained by the network at all.
This might not be the behavior we want.
Sequence models are central to NLP: they are models where there is some sort of dependence through time between your inputs.
The classical example of a sequence model is the Hidden Markov Model for part-of-speech tagging.  Another example is the conditional random field.

A recurrent neural network is a network that maintains some kind of state.
For example, its output could be used as part of the next input, so that information can propogate along as the network passes over the sequence.
In the case of an LSTM, for each element in the sequence, there is a corresponding *hidden state* $h_t$, which in principle can contain information from arbitrary points earlier in the sequence.
We can use the hidden state to predict words in a language model, part-of-speech tags, and a myriad of other things.

### LSTM's in Pytorch

<font color=red>Before getting to the example, note a few things.
Pytorch's LSTM expects all of its inputs to be 3D tensors.
The semantics of the axes of these tensors is important.
The first axis is the sequence itself, the second indexes instances in the mini-batch, and the third indexes elements of the input.</font>
We haven't discussed mini-batching, so lets just ignore that and assume we will always have just 1 dimension on the second axis.
If we want to run the sequence model over the sentence "The cow jumped", our input should look like
$$ 
\begin{bmatrix}
\overbrace{q_\text{The}}^\text{row vector} \\
q_\text{cow} \\
q_\text{jumped}
\end{bmatrix}
$$
Except remember there is an additional 2nd dimension with size 1.

In addition, you could go through the sequence one at a time, in which case the 1st axis will have size 1 also.

Let's see a quick example.

In [37]:
lstm = nn.LSTM(3, 3) # Input dim is 3, output (hidden) dim is 3

# From PyTorch documentation:
# Inputs: input, (h_0, c_0)
# input of shape (seq_len, batch, input_size)
# h_0 of shape (num_layers * num_directions, batch, hidden_size)
# c_0 of shape (num_layers * num_directions, batch, hidden_size)
# Also, check the Outputs description from the documentation.

inputs = [torch.randn(1, 3) for _ in range(5)]

# initialize the hidden state.  
hidden = (torch.randn(1,1,3), torch.randn((1,1,3)))

for i in inputs:
    # Step through the sequence one element at a time.
    # after each step, hidden contains the hidden state.
    out, hidden = lstm(i.view(1,1,-1), hidden)
    print(out)
    print(hidden)

print("-----------------------------------------------------------------------------------")
    
# alternatively, we can do the entire sequence all at once.
# the first value returned by LSTM is all of the hidden states throughout the sequence.
# the second is just the most recent hidden state and cell state (compare the last slice of "out" with "hidden[0]" below,
# they are the same)
# The reason for this is that:
# "out" will give you access to all hidden states in the sequence
# "hidden" will allow you to continue the sequence and backpropagate, by passing it as an argument
# to the lstm at a later time
inputs = torch.cat(inputs).view(len(inputs), 1, -1) # Add the extra 2nd dimension
hidden = (torch.randn(1,1,3), torch.randn((1,1,3))) # clean out hidden state
out, hidden = lstm(inputs, hidden)
print(out)
print(hidden)

tensor([[[-0.0696, -0.1333,  0.0565]]])
(tensor([[[-0.0696, -0.1333,  0.0565]]]), tensor([[[-0.1495, -0.3605,  0.1222]]]))
tensor(1.00000e-02 *
       [[[ 4.9250, -6.9015,  4.2336]]])
(tensor(1.00000e-02 *
       [[[ 4.9250, -6.9015,  4.2336]]]), tensor([[[ 0.1036, -0.1555,  0.1805]]]))
tensor([[[ 0.2914,  0.1347,  0.0594]]])
(tensor([[[ 0.2914,  0.1347,  0.0594]]]), tensor([[[ 0.3980,  0.2017,  0.2042]]]))
tensor([[[ 0.1470,  0.1463,  0.2308]]])
(tensor([[[ 0.1470,  0.1463,  0.2308]]]), tensor([[[ 0.2194,  0.2734,  0.3968]]]))
tensor([[[ 0.1608,  0.1230,  0.1154]]])
(tensor([[[ 0.1608,  0.1230,  0.1154]]]), tensor([[[ 0.2614,  0.2409,  0.3168]]]))
-----------------------------------------------------------------------------------
tensor([[[-0.3066,  0.1390, -0.2117]],

        [[-0.1106,  0.0358, -0.0125]],

        [[ 0.1419,  0.1599,  0.0351]],

        [[ 0.0455,  0.1463,  0.1877]],

        [[ 0.0957,  0.1148,  0.1037]]])
(tensor([[[ 0.0957,  0.1148,  0.1037]]]), tensor([[[ 0.1544

### Example: An LSTM for Part-of-Speech Tagging

In this section, we will use an LSTM to get part of speech tags.
We will not use Viterbi or Forward-Backward or anything like that, but as a (challenging) exercise to the reader, think about how Viterbi could be used after you have seen what is going on.

The model is as follows: let our input sentence be $w_1, \dots, w_M$, where $w_i \in V$, our vocab.
Also, let $T$ be our tag set, and $y_i$ the tag of word $w_i$.  Denote our prediction of the tag of word $w_i$ by $\hat{y}_i$.

This is a structure prediction, model, where our output is a sequence $\hat{y}_1, \dots, \hat{y}_M$, where $\hat{y}_i \in T$.

To do the prediction, pass an LSTM over the sentence.  Denote the hidden state at timestep $i$ as $h_i$.  Also, assign each tag a unique index (like how we had word_to_ix in the word embeddings section).
Then our prediction rule for $\hat{y}_i$ is
$$ \hat{y}_i = \text{argmax}_j \  (\log \text{Softmax}(Ah_i + b))_j $$
That is, take the log softmax of the affine map of the hidden state, and the predicted tag is the tag that has the maximum value in this vector.  Note this implies immediately that the dimensionality of the target space of $A$ is $|T|$.

In [115]:
def prepare_sequence(seq, to_ix):
    idxs = list(map(lambda w: to_ix[w], seq))
    tensor = torch.LongTensor(idxs)
    return tensor

In [116]:
training_data = [
    ("The dog ate the apple".split(), ["DET", "NN", "V", "DET", "NN"]),
    ("Everybody read that book".split(), ["NN", "V", "DET", "NN"])
]
word_to_ix = {}
for sent, tags in training_data:
    for word in sent:
        if word not in word_to_ix:
            word_to_ix[word] = len(word_to_ix)
print(word_to_ix)
tag_to_ix = {"DET": 0, "NN": 1, "V": 2}

# These will usually be more like 32 or 64 dimensional.
# We will keep them small, so we can see how the weights change as we train.
EMBEDDING_DIM = 6
HIDDEN_DIM = 6

{'apple': 4, 'ate': 2, 'the': 3, 'book': 8, 'read': 6, 'Everybody': 5, 'that': 7, 'The': 0, 'dog': 1}


In [117]:
class LSTMTagger(nn.Module):
    
    def __init__(self, embedding_dim, hidden_dim, vocab_size, tagset_size):
        super(LSTMTagger, self).__init__()
        self.hidden_dim = hidden_dim
        
        self.word_embeddings = nn.Embedding(vocab_size, embedding_dim)
        
        # The LSTM takes word embeddings as inputs, and outputs hidden states
        # with dimensionality hidden_dim.
        self.lstm = nn.LSTM(embedding_dim, hidden_dim)
        
        # The linear layer that maps from hidden state space to tag space
        self.hidden2tag = nn.Linear(hidden_dim, tagset_size)
        self.hidden = self.init_hidden()
        
    def init_hidden(self):
        # Before we've done anything, we dont have any hidden state.
        # Refer to the Pytorch documentation to see exactly why they have this dimensionality.
        # The axes semantics are (num_layers * num_directions, minibatch_size, hidden_dim)
        return (torch.zeros(1, 1, self.hidden_dim),
                torch.zeros(1, 1, self.hidden_dim))
        
    def forward(self, sentence):
        embeds = self.word_embeddings(sentence)
        lstm_out, self.hidden = self.lstm(embeds.view(len(sentence), 1, -1), self.hidden)
        tag_space = self.hidden2tag(lstm_out.view(len(sentence), -1))
        tag_scores = F.log_softmax(tag_space, dim=1)
        return tag_scores

In [118]:
model = LSTMTagger(EMBEDDING_DIM, HIDDEN_DIM, len(word_to_ix), len(tag_to_ix))
loss_function = nn.NLLLoss()
optimizer = optim.SGD(model.parameters(), lr=0.1)

In [119]:
# See what the scores are before training
# Note that element i,j of the output is the score for tag j for word i.
inputs = prepare_sequence(training_data[0][0], word_to_ix)
print(inputs)
tag_scores = model(inputs)
print(tag_scores)

tensor([ 0,  1,  2,  3,  4])
tensor([[-1.3530, -1.1678, -0.8428],
        [-1.3426, -1.1253, -0.8812],
        [-1.3615, -1.1463, -0.8535],
        [-1.3591, -1.1393, -0.8602],
        [-1.3369, -1.1440, -0.8704]])


In [120]:
for epoch in range(300): # again, normally you would NOT do 300 epochs, it is toy data
    for sentence, tags in training_data:
        # Step 1. Remember that Pytorch accumulates gradients.  We need to clear them out
        # before each instance
        model.zero_grad()
        
        # Also, we need to clear out the hidden state of the LSTM, detaching it from its
        # history on the last instance.
        model.hidden = model.init_hidden()
    
        # Step 2. Get our inputs ready for the network, that is, turn them into Variables
        # of word indices.
        sentence_in = prepare_sequence(sentence, word_to_ix)
        targets = prepare_sequence(tags, tag_to_ix)
    
        # Step 3. Run our forward pass.
        tag_scores = model(sentence_in)
    
        # Step 4. Compute the loss, gradients, and update the parameters by calling
        # optimizer.step()
        loss = loss_function(tag_scores, targets)
        loss.backward()
        optimizer.step()
    print("Finished epoch {}, with loss: {:.4f}".format(epoch, loss))

Finished epoch 0, with loss: 1.1345
Finished epoch 1, with loss: 1.1213
Finished epoch 2, with loss: 1.1098
Finished epoch 3, with loss: 1.0996
Finished epoch 4, with loss: 1.0906
Finished epoch 5, with loss: 1.0826
Finished epoch 6, with loss: 1.0754
Finished epoch 7, with loss: 1.0689
Finished epoch 8, with loss: 1.0631
Finished epoch 9, with loss: 1.0577
Finished epoch 10, with loss: 1.0528
Finished epoch 11, with loss: 1.0482
Finished epoch 12, with loss: 1.0439
Finished epoch 13, with loss: 1.0398
Finished epoch 14, with loss: 1.0359
Finished epoch 15, with loss: 1.0322
Finished epoch 16, with loss: 1.0286
Finished epoch 17, with loss: 1.0251
Finished epoch 18, with loss: 1.0217
Finished epoch 19, with loss: 1.0183
Finished epoch 20, with loss: 1.0150
Finished epoch 21, with loss: 1.0117
Finished epoch 22, with loss: 1.0083
Finished epoch 23, with loss: 1.0050
Finished epoch 24, with loss: 1.0017
Finished epoch 25, with loss: 0.9983
Finished epoch 26, with loss: 0.9949
Finished ep

Finished epoch 237, with loss: 0.0914
Finished epoch 238, with loss: 0.0904
Finished epoch 239, with loss: 0.0894
Finished epoch 240, with loss: 0.0884
Finished epoch 241, with loss: 0.0875
Finished epoch 242, with loss: 0.0865
Finished epoch 243, with loss: 0.0856
Finished epoch 244, with loss: 0.0847
Finished epoch 245, with loss: 0.0838
Finished epoch 246, with loss: 0.0829
Finished epoch 247, with loss: 0.0821
Finished epoch 248, with loss: 0.0812
Finished epoch 249, with loss: 0.0804
Finished epoch 250, with loss: 0.0796
Finished epoch 251, with loss: 0.0788
Finished epoch 252, with loss: 0.0780
Finished epoch 253, with loss: 0.0772
Finished epoch 254, with loss: 0.0765
Finished epoch 255, with loss: 0.0757
Finished epoch 256, with loss: 0.0750
Finished epoch 257, with loss: 0.0743
Finished epoch 258, with loss: 0.0736
Finished epoch 259, with loss: 0.0729
Finished epoch 260, with loss: 0.0722
Finished epoch 261, with loss: 0.0715
Finished epoch 262, with loss: 0.0708
Finished epo

In [123]:
# Don't forget this! That's why I think it should be done at the beginning of forward.
model.hidden = model.init_hidden()

# See what the scores are after training
inputs = prepare_sequence(training_data[0][0], word_to_ix)
tag_scores = model(inputs)
# The sentence is "the dog ate the apple".  i,j corresponds to score for tag j for word i.
# The predicted tag is the maximum scoring tag.
# Here, we can see the predicted sequence below is 0 1 2 0 1
# since 0 is index of the maximum value of row 1,
# 1 is the index of maximum value of row 2, etc.
# Which is DET NOUN VERB DET NOUN, the correct sequence!
print(tag_scores)

tensor([[-0.0699, -2.9567, -4.1639],
        [-3.6291, -0.0812, -2.9673],
        [-4.8594, -2.9503, -0.0620],
        [-0.0441, -3.3160, -4.9904],
        [-3.6265, -0.0436, -4.1290]])


### Exercise: Augmenting the LSTM part-of-speech tagger with character-level features
In the example above, each word had an embedding, which served as the inputs to our sequence model.
Let's augment the word embeddings with a representation derived from the characters of the word.
We expect that this should help significantly, since character-level information like affixes have
a large bearing on part-of-speech.  For example, words with the affix *-ly* are almost always tagged as adverbs in English.

Do do this, let $c_w$ be the character-level representation of word $w$.  Let $x_w$ be the word embedding as before.
Then the input to our sequence model is the concatenation of $x_w$ and $c_w$.  So if $x_w$ has dimension 5, and $c_w$ dimension 3, then our LSTM should accept an input of dimension 8.

To get the character level representation, do an LSTM over the characters of a word, and let $c_w$ be the final hidden state of this LSTM.
Hints:
* There are going to be two LSTM's in your new model.  The original one that outputs POS tag scores, and the new one that outputs a character-level representation of each word. <br>
<font color=red>?: Is it better to train them together or pass the trained CharLSTM? Is it even possible to use a trained model within another one without training the former (I guess I could set requires_grad to False; I'm gonna ask James about this anyways)?</font>
* To do a sequence model over characters, you will have to embed characters.  The character embeddings will be the input to the character LSTM.

In [124]:
char_to_ix = {}
for sent, tags in training_data:
    for word in sent:
        for char in word:
            if char not in char_to_ix:
                char_to_ix[char] = len(char_to_ix)
print(char_to_ix)

{'a': 6, 'e': 2, 'o': 4, 'l': 9, 'k': 15, 'v': 11, 'p': 8, 'g': 5, 'h': 1, 'd': 3, 'r': 12, 'E': 10, 'b': 14, 't': 7, 'T': 0, 'y': 13}


In [125]:
class CharLSTM(nn.Module):

    def __init__(self, embedding_dim, hidden_dim, char_size):
        super(CharLSTM, self).__init__()
        self.hidden_dim = hidden_dim
        
        self.char_embeddings = nn.Embedding(char_size, embedding_dim)
        
        # The LSTM takes char embeddings as inputs, and outputs hidden states
        # with dimensionality hidden_dim.
        self.lstm = nn.LSTM(embedding_dim, hidden_dim)

        self.hidden = self.init_hidden()
        
    def init_hidden(self):
        # Before we've done anything, we dont have any hidden state.
        # Refer to the Pytorch documentation to see exactly why they have this dimensionality.
        # The axes semantics are (num_layers * num_directions, minibatch_size, hidden_dim)
        return (torch.zeros(1, 1, self.hidden_dim),
                torch.zeros(1, 1, self.hidden_dim))
        
    def forward(self, word):
        self.hidden = self.init_hidden()  # I'm gonna do this at the beginning of forward.
        embeds = self.char_embeddings(word)
        lstm_out, self.hidden = self.lstm(embeds.view(len(word), 1, -1), self.hidden)
        return self.hidden[0]

In [126]:
class LSTMTaggerWithCharInfo(nn.Module):
    
    def __init__(self, embedding_dim, hidden_dim, vocab_size, tagset_size, 
                 char_embedding_dim, char_hidden_dim, char_size):
        super(LSTMTaggerWithCharInfo, self).__init__()
        self.hidden_dim = hidden_dim
        
        self.char_representations = CharLSTM(char_embedding_dim, char_hidden_dim, char_size)
        self.word_embeddings = nn.Embedding(vocab_size, embedding_dim)
        
        # The LSTM takes word embeddings as inputs, and outputs hidden states
        # with dimensionality hidden_dim.
        self.lstm = nn.LSTM(embedding_dim + char_embedding_dim, hidden_dim)
        
        # The linear layer that maps from hidden state space to tag space
        self.hidden2tag = nn.Linear(hidden_dim, tagset_size)
        self.hidden = self.init_hidden()
        
    def init_hidden(self):
        # Before we've done anything, we dont have any hidden state.
        # Refer to the Pytorch documentation to see exactly why they have this dimensionality.
        # The axes semantics are (num_layers * num_directions, minibatch_size, hidden_dim)
        return (torch.zeros(1, 1, self.hidden_dim),
                torch.zeros(1, 1, self.hidden_dim))
        
    def forward(self, sentence, words):
        char_info = torch.stack([self.char_representations(word).view(-1) for word in words])
        embeds = self.word_embeddings(sentence)
        lstm_input = torch.cat((embeds, char_info), 1)
        lstm_out, self.hidden = self.lstm(lstm_input.view(len(sentence), 1, -1), self.hidden)
        tag_space = self.hidden2tag(lstm_out.view(len(sentence), -1))
        tag_scores = F.log_softmax(tag_space, dim=1)
        return tag_scores

In [127]:
model = LSTMTaggerWithCharInfo(6, 6, len(word_to_ix), len(tag_to_ix), 3, 3, len(char_to_ix))
loss_function = nn.NLLLoss()
optimizer = optim.SGD(model.parameters(), lr=0.1)

for epoch in range(300): # again, normally you would NOT do 300 epochs, it is toy data
    for sentence, tags in training_data:
        # Step 1. Remember that Pytorch accumulates gradients.  We need to clear them out
        # before each instance
        model.zero_grad()
        
        # Also, we need to clear out the hidden state of the LSTM, detaching it from its
        # history on the last instance.
        model.hidden = model.init_hidden()
    
        # Step 2. Get our inputs ready for the network, that is, turn them into Variables
        # of word indices.
        sentence_in = prepare_sequence(sentence, word_to_ix)
        words_in = [prepare_sequence(word, char_to_ix) for word in sentence]
        targets = prepare_sequence(tags, tag_to_ix)
    
        # Step 3. Run our forward pass.
        tag_scores = model(sentence_in, words_in)
    
        # Step 4. Compute the loss, gradients, and update the parameters by calling
        # optimizer.step()
        loss = loss_function(tag_scores, targets)
        loss.backward()
        optimizer.step()
    print("Finished epoch {}, with loss: {:.4f}".format(epoch, loss))

Finished epoch 0, with loss: 1.1162
Finished epoch 1, with loss: 1.1092
Finished epoch 2, with loss: 1.1027
Finished epoch 3, with loss: 1.0969
Finished epoch 4, with loss: 1.0915
Finished epoch 5, with loss: 1.0865
Finished epoch 6, with loss: 1.0819
Finished epoch 7, with loss: 1.0775
Finished epoch 8, with loss: 1.0735
Finished epoch 9, with loss: 1.0697
Finished epoch 10, with loss: 1.0661
Finished epoch 11, with loss: 1.0626
Finished epoch 12, with loss: 1.0593
Finished epoch 13, with loss: 1.0562
Finished epoch 14, with loss: 1.0531
Finished epoch 15, with loss: 1.0502
Finished epoch 16, with loss: 1.0473
Finished epoch 17, with loss: 1.0445
Finished epoch 18, with loss: 1.0417
Finished epoch 19, with loss: 1.0390
Finished epoch 20, with loss: 1.0363
Finished epoch 21, with loss: 1.0336
Finished epoch 22, with loss: 1.0309
Finished epoch 23, with loss: 1.0282
Finished epoch 24, with loss: 1.0255
Finished epoch 25, with loss: 1.0227
Finished epoch 26, with loss: 1.0199
Finished ep

Finished epoch 222, with loss: 0.0793
Finished epoch 223, with loss: 0.0783
Finished epoch 224, with loss: 0.0773
Finished epoch 225, with loss: 0.0763
Finished epoch 226, with loss: 0.0754
Finished epoch 227, with loss: 0.0745
Finished epoch 228, with loss: 0.0736
Finished epoch 229, with loss: 0.0727
Finished epoch 230, with loss: 0.0718
Finished epoch 231, with loss: 0.0710
Finished epoch 232, with loss: 0.0702
Finished epoch 233, with loss: 0.0693
Finished epoch 234, with loss: 0.0685
Finished epoch 235, with loss: 0.0678
Finished epoch 236, with loss: 0.0670
Finished epoch 237, with loss: 0.0662
Finished epoch 238, with loss: 0.0655
Finished epoch 239, with loss: 0.0648
Finished epoch 240, with loss: 0.0641
Finished epoch 241, with loss: 0.0634
Finished epoch 242, with loss: 0.0627
Finished epoch 243, with loss: 0.0620
Finished epoch 244, with loss: 0.0614
Finished epoch 245, with loss: 0.0607
Finished epoch 246, with loss: 0.0601
Finished epoch 247, with loss: 0.0595
Finished epo

In [129]:
# Don't forget this! That's why I think it should be done at the beginning of forward.
model.hidden = model.init_hidden()

# See what the scores are after training
sentence_in = prepare_sequence(training_data[0][0], word_to_ix)
words_in = [prepare_sequence(word, char_to_ix) for word in training_data[0][0]]
tag_scores = model(sentence_in, words_in)
# The sentence is "the dog ate the apple".  i,j corresponds to score for tag j for word i.
# The predicted tag is the maximum scoring tag.
# Here, we can see the predicted sequence below is 0 1 2 0 1
# since 0 is index of the maximum value of row 1,
# 1 is the index of maximum value of row 2, etc.
# Which is DET NOUN VERB DET NOUN, the correct sequence!
print(tag_scores)

tensor([[-0.0706, -3.2247, -3.5612],
        [-4.0338, -0.0612, -3.1777],
        [-2.8293, -3.5798, -0.0909],
        [-0.0287, -5.4607, -3.7299],
        [-4.4610, -0.0219, -4.5988]])


Previous results for comparison: <br>
`tensor([[-0.0699, -2.9567, -4.1639],
         [-3.6291, -0.0812, -2.9673],
         [-4.8594, -2.9503, -0.0620],
         [-0.0441, -3.3160, -4.9904],
         [-3.6265, -0.0436, -4.1290]])`

# 8. Advanced: Dynamic Toolkits, Dynamic Programming, and the BiLSTM-CRF

### Dyanmic versus Static Deep Learning Toolkits

Pytorch is a *dynamic* neural network kit.  Another example of a dynamic kit is [Dynet](https://github.com/clab/dynet) (I mention this because working with Pytorch and Dynet is similar.  If you see an example in Dynet, it will probably help you implement it in Pytorch).  The opposite is the *static* tool kit, which includes Theano, Keras, TensorFlow, etc.
The core difference is the following:
* In a static toolkit, you define a computation graph once, compile it, and then stream instances to it.
* In a dynamic toolkit, you define a computation graph *for each instance*.  It is never compiled and is executed on-the-fly

Without a lot of experience, it is difficult to appreciate the difference.
One example is to suppose we want to build a deep constituent parser.
Suppose our model involves roughly the following steps:
* We build the tree bottom up
* Tag the root nodes (the words of the sentence)
* From there, use a neural network and the embeddings of the words
to find combinations that form constituents.  Whenever you form a new constituent,
use some sort of technique to get an embedding of the constituent.
<font color=red>In this case, our network architecture will depend completely on the input sentence.</font>
In the sentence "The green cat scratched the wall", at some point in the model, we will want to combine
the span $(i,j,r) = (1, 3, \text{NP})$ (that is, an NP constituent spans word 1 to word 3, in this case "The green cat").

However, another sentence might be "Somewhere, the big fat cat scratched the wall".  In this sentence, we will want to form the constituent $(2, 4, NP)$ at some point.
The constituents we will want to form will depend on the instance.  If we just compile the computation graph once, as in a static toolkit, it will be exceptionally difficult or impossible to program this logic.  In a dynamic toolkit though, there isn't just 1 pre-defined computation graph.  There can be a new computation graph for each instance, so this problem goes away.

Dynamic toolkits also have the advantage of being easier to debug and the code more closely resembling the host language (by that I mean that Pytorch and Dynet look more like actual Python code than Keras or Theano).

I mention this distinction here, because the exercise in this section is to implement a model which closely resembles structure perceptron, and I believe this model would be difficult to implement in a static toolkit.  I think that the advantage of dynamic toolkits for linguistic structure prediction cannot be overstated.

### Bi-LSTM Conditional Random Field Discussion

For this section, we will see a full, complicated example of a Bi-LSTM Conditional Random Field for named-entity recognition.  The LSTM tagger above is typically sufficient for part-of-speech tagging, but a sequence model like the CRF is really essential for strong performance on NER.  Familiarity with CRF's is assumed.  Although this name sounds scary, all the model is is a CRF but where an LSTM provides the features.  This is an advanced model though, far more complicated than any earlier model in this tutorial.  If you want to skip it, that is fine.  To see if you're ready, see if you can:

* Write the recurrence for the viterbi variable at step i for tag k.
* Modify the above recurrence to compute the forward variables instead.
* Modify again the above recurrence to compute the forward variables in log-space (hint: log-sum-exp)

If you can do those three things, you should be able to understand the code below.
Recall that the CRF computes a conditional probability.  Let $y$ be a tag sequence and $x$ an input sequence of words.  Then we compute
$$ P(y|x) = \frac{\exp{(\text{Score}(x, y)})}{\sum_{y'} \exp{(\text{Score}(x, y')})} $$

Where the score is determined by defining some log potentials $\log \psi_i(x,y)$ such that
$$ \text{Score}(x,y) = \sum_i \log \psi_i(x,y) $$
To make the partition function tractable, the potentials must look only at local features.

In the Bi-LSTM CRF, we define two kinds of potentials: emission and transition.  The emission potential for the word at index $i$ comes from the hidden state of the Bi-LSTM at timestep $i$.  The transition scores are stored in a $|T|x|T|$ matrix $\textbf{P}$, where $T$ is the tag set.  In my implementation, $\textbf{P}_{j,k}$ is the score of transitioning to tag $j$ from tag $k$.  So:

$$ \text{Score}(x,y) = \sum_i \log \psi_\text{EMIT}(y_i \rightarrow x_i) + \log \psi_\text{TRANS}(y_{i-1} \rightarrow y_i) $$
$$ = \sum_i h_i[y_i] + \textbf{P}_{y_i, y_{i-1}} $$
where in this second expression, we think of the tags as being assigned unique non-negative indices.

If the above discussion was too brief, you can check out [this](http://www.cs.columbia.edu/%7Emcollins/crf.pdf) write up from Michael Collins on CRFs.

### The Forward Algorithm in Log-Space and the Log-Sum-Exp Trick

As hinted at above, computing the forward variables requires using a log-sum-exp.  I want to explain why, since it was a little confusing to me at first, and many resources just present the forward algorithm in potential space.  The recurrence for the forward variable at the $i$'th word for the tag $j$, $\alpha_i(j)$, is
$$ \alpha_i(j) = \sum_{j' \in T} \psi_\text{EMIT}(j \rightarrow i) \times \psi_\text{TRANS}(j' \rightarrow j) \times \alpha_{i-1}(j') $$

This is numerically unstable, and underflow is likely.  It is also inconvenient to work with proper non-negative potentials in our model.  We instead want to compute $\log \alpha_i(j)$.  What we need to do is to multiply the potentials, which corresponds to adding log potentials.  Then, we have to sum over tags, but what is the corresponding operation to summing over tags in log space?  It is not clear.  Instead, we need to transform out of log-space, take the product of potentials, do the sum over tags, and then transform back to log space.  This is broken down in the revised recurrence below:

$$ \log \alpha_i(j) = \log \overbrace{\sum_{j' \in T} \exp{(\log \psi_\text{EMIT}(j \rightarrow i) + \log \psi_\text{TRANS}(j' \rightarrow j) + \log \alpha_{i-1}(j'))}}^\text{transform out of log-space and compute forward variable} $$

If you carry out elementary exponential / logarithm identities in the stuff under the overbrace above, you will see that it computes the same thing as the first recurrence, then just takes the logarithm.  Log-sum-exp appears a fair bit in machine learning, and there is a [well-known trick](https://en.wikipedia.org/wiki/LogSumExp) to computing it in a numerically stable way.  I use this trick in my log_sum_exp function below (I don't think Pytorch provides this function in its library).

### Implementation Notes

The example below implements the forward algorithm in log space to compute the partition function, and the viterbi algorithm to decode.  Backpropagation will compute the gradients automatically for us.  We don't have to do anything by hand.

The implementation is not optimized.  If you understand what is going on, you'll probably quickly see that iterating over the next tag in the forward algorithm could probably be done in one big operation.  I wanted to code to be more readable.  If you want to make the relevant change, you could probably use this tagger for real tasks.

### Example: Bidirectional LSTM Conditional Random Field for Named-Entity Recognition

In [47]:
# Helper functions to make the code more readable.
def to_scalar(var):
    # returns a python float
    return var.view(-1).data.tolist()[0]

def argmax(vec):
    # return the argmax as a python int
    _, idx = torch.max(vec, 1)
    return to_scalar(idx)

# Compute log sum exp in a numerically stable way for the forward algorithm
def log_sum_exp(vec):
    max_score = vec[0, argmax(vec)]
    max_score_broadcast = max_score.view(1, -1).expand(1, vec.size()[1])
    return max_score + torch.log(torch.sum(torch.exp(vec - max_score_broadcast)))
    

class BiLSTM_CRF(nn.Module):
    
    def __init__(self, vocab_size, tag_to_ix, embedding_dim, hidden_dim):
        super(BiLSTM_CRF, self).__init__()
        self.embedding_dim = embedding_dim
        self.hidden_dim = hidden_dim
        self.vocab_size = vocab_size
        self.tag_to_ix = tag_to_ix
        self.tagset_size = len(tag_to_ix)
        
        self.word_embeds = nn.Embedding(vocab_size, embedding_dim)
        self.lstm = nn.LSTM(embedding_dim, hidden_dim/2, num_layers=1, bidirectional=True)
        
        # Maps the output of the LSTM into tag space.
        self.hidden2tag = nn.Linear(hidden_dim, self.tagset_size)
        
        # Matrix of transition parameters.  Entry i,j is the score of transitioning *to* i *from* j.
        self.transitions = nn.Parameter(torch.randn(self.tagset_size, self.tagset_size))
        
        # These two statements enforce the constraint that we never transfer *to* the start tag,
        # and we never transfer *from* the stop tag (the model would probably learn this anyway,
        # so this enforcement is likely unimportant)
        self.transitions.data[tag_to_ix[START_TAG], :] = -10000
        self.transitions.data[:, tag_to_ix[STOP_TAG]] = -10000
        
        self.hidden = self.init_hidden()
        
    def init_hidden(self):
        return ( autograd.Variable( torch.randn(2, 1, self.hidden_dim)),
                 autograd.Variable( torch.randn(2, 1, self.hidden_dim)) )
    
    
    def _forward_alg(self, feats):
        # Do the forward algorithm to compute the partition function
        init_alphas = torch.Tensor(1, self.tagset_size).fill_(-10000.)
        # START_TAG has all of the score.
        init_alphas[0][self.tag_to_ix[START_TAG]] = 0.
        
        # Wrap in a variable so that we will get automatic backprop
        forward_var = autograd.Variable(init_alphas)
        
        # Iterate through the sentence
        for feat in feats:
            alphas_t = [] # The forward variables at this timestep
            for next_tag in xrange(self.tagset_size):
                # broadcast the emission score: it is the same regardless of the previous tag
                emit_score = feat[next_tag].view(1, -1).expand(1, self.tagset_size)
                # the ith entry of trans_score is the score of transitioning to next_tag from i
                trans_score = self.transitions[next_tag].view(1, -1)
                # The ith entry of next_tag_var is the value for the edge (i -> next_tag)
                # before we do log-sum-exp
                next_tag_var = forward_var + trans_score + emit_score
                # The forward variable for this tag is log-sum-exp of all the scores.
                alphas_t.append(log_sum_exp(next_tag_var))
            forward_var = torch.cat(alphas_t).view(1, -1)
        terminal_var = forward_var + self.transitions[self.tag_to_ix[STOP_TAG]]
        alpha = log_sum_exp(terminal_var)
        return alpha
        
    def _get_lstm_features(self, sentence):
        self.hidden = self.init_hidden()
        embeds = self.word_embeds(sentence).view(len(sentence), 1, -1)
        lstm_out, self.hidden = self.lstm(embeds)
        lstm_out = lstm_out.view(len(sentence), self.hidden_dim)
        lstm_feats = self.hidden2tag(lstm_out)
        return lstm_feats
        
    def _score_sentence(self, feats, tags):
        # Gives the score of a provided tag sequence
        score = autograd.Variable( torch.Tensor([0]) )
        tags = torch.cat( [torch.LongTensor([self.tag_to_ix[START_TAG]]), tags] )
        for i, feat in enumerate(feats):
            score = score + self.transitions[tags[i+1], tags[i]] + feat[tags[i+1]]
        score = score + self.transitions[self.tag_to_ix[STOP_TAG], tags[-1]]
        return score
    
    def _viterbi_decode(self, feats):
        backpointers = []
        
        # Initialize the viterbi variables in log space
        init_vvars = torch.Tensor(1, self.tagset_size).fill_(-10000.)
        init_vvars[0][self.tag_to_ix[START_TAG]] = 0
        
        # forward_var at step i holds the viterbi variables for step i-1 
        forward_var = autograd.Variable(init_vvars)
        for feat in feats:
            bptrs_t = [] # holds the backpointers for this step
            viterbivars_t = [] # holds the viterbi variables for this step
            
            for next_tag in range(self.tagset_size):
                # next_tag_var[i] holds the viterbi variable for tag i at the previous step,
                # plus the score of transitioning from tag i to next_tag.
                # We don't include the emission scores here because the max
                # does not depend on them (we add them in below)
                next_tag_var = forward_var + self.transitions[next_tag]
                best_tag_id = argmax(next_tag_var)
                bptrs_t.append(best_tag_id)
                viterbivars_t.append(next_tag_var[0][best_tag_id])
            # Now add in the emission scores, and assign forward_var to the set
            # of viterbi variables we just computed
            forward_var = (torch.cat(viterbivars_t) + feat).view(1, -1)
            backpointers.append(bptrs_t)
        
        # Transition to STOP_TAG
        terminal_var = forward_var + self.transitions[self.tag_to_ix[STOP_TAG]]
        best_tag_id = argmax(terminal_var)
        path_score = terminal_var[0][best_tag_id]
        
        # Follow the back pointers to decode the best path.
        best_path = [best_tag_id]
        for bptrs_t in reversed(backpointers):
            best_tag_id = bptrs_t[best_tag_id]
            best_path.append(best_tag_id)
        # Pop off the start tag (we dont want to return that to the caller)
        start = best_path.pop()
        assert start == self.tag_to_ix[START_TAG] # Sanity check
        best_path.reverse()
        return path_score, best_path
        
    def neg_log_likelihood(self, sentence, tags):
        self.hidden = self.init_hidden()
        feats = self._get_lstm_features(sentence)
        forward_score = self._forward_alg(feats)
        gold_score = self._score_sentence(feats, tags)
        return forward_score - gold_score
        
    def forward(self, sentence): # dont confuse this with _forward_alg above.
        self.hidden = self.init_hidden()
        # Get the emission scores from the BiLSTM
        lstm_feats = self._get_lstm_features(sentence)
        
        # Find the best path, given the features.
        score, tag_seq = self._viterbi_decode(lstm_feats)
        return score, tag_seq


In [48]:
START_TAG = "<START>"
STOP_TAG = "<STOP>"
EMBEDDING_DIM = 5
HIDDEN_DIM = 4

# Make up some training data
training_data = [ (
    "the wall street journal reported today that apple corporation made money".split(),
    "B I I I O O O B I O O".split()
), (
    "georgia tech is a university in georgia".split(),
    "B I O O O O B".split()
) ]

word_to_ix = {}
for sentence, tags in training_data:
    for word in sentence:
        if word not in word_to_ix:
            word_to_ix[word] = len(word_to_ix)
            
tag_to_ix = { "B": 0, "I": 1, "O": 2, START_TAG: 3, STOP_TAG: 4 }

In [49]:
model = BiLSTM_CRF(len(word_to_ix), tag_to_ix, EMBEDDING_DIM, HIDDEN_DIM)
optimizer = optim.SGD(model.parameters(), lr=0.01, weight_decay=1e-4)

In [50]:
# Check predictions before training
precheck_sent = prepare_sequence(training_data[0][0], word_to_ix)
precheck_tags = torch.LongTensor([ tag_to_ix[t] for t in training_data[0][1] ])
print model(precheck_sent)

(Variable containing:
 13.2216
[torch.FloatTensor of size 1]
, [2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 0])


In [51]:
# Make sure prepare_sequence from earlier in the LSTM section is loaded
for epoch in xrange(300): # again, normally you would NOT do 300 epochs, it is toy data
    for sentence, tags in training_data:
        # Step 1. Remember that Pytorch accumulates gradients.  We need to clear them out
        # before each instance
        model.zero_grad()
    
        # Step 2. Get our inputs ready for the network, that is, turn them into Variables
        # of word indices.
        sentence_in = prepare_sequence(sentence, word_to_ix)
        targets = torch.LongTensor([ tag_to_ix[t] for t in tags ])
    
        # Step 3. Run our forward pass.
        neg_log_likelihood = model.neg_log_likelihood(sentence_in, targets)
    
        # Step 4. Compute the loss, gradients, and update the parameters by calling
        # optimizer.step()
        neg_log_likelihood.backward()
        optimizer.step()

In [52]:
# Check predictions after training
precheck_sent = prepare_sequence(training_data[0][0], word_to_ix)
print model(precheck_sent)
# We got it!

(Variable containing:
 31.9680
[torch.FloatTensor of size 1]
, [0, 1, 1, 1, 2, 2, 2, 0, 1, 2, 2])


### Exercise: A new loss function for discriminative tagging
It wasn't really necessary for us to create a computation graph when doing decoding, since we do not backpropagate from the viterbi path score.  Since we have it anyway, try training the tagger where the loss function is the difference between the Viterbi path score and the score of the gold-standard path.  It should be clear that this function is non-negative and 0 when the predicted tag sequence is the correct tag sequence.  This is essentially *structured perceptron*.

This modification should be short, since Viterbi and score_sentence are already implemented.  This is an example of the shape of the computation graph *depending on the training instance*.  Although I haven't tried implementing this in a static toolkit, I imagine that it is possible but much less straightforward.

Pick up some real data and do a comparison!